Unhelpful "not a struct" error when typoing a type name when opening an impl block and using `Self`

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

Given the following code: <!-- Please provide a link to play.rust-lang.org -->

```rust
extern crate regex; // 1.5.4

use std::iter::Enumerate;

pub struct CaptureIndices<'a, 'b> {
    name: &'b [u8],
    capture_names: Enumerate<regex::CaptureNames<'a>>
}

impl<'a, 'b> CapturesIndicies<'a, 'b> {
    pub(crate) fn with_name_and_iter(name: &'b [u8], iter: regex::CaptureNames<'a>) -> Self {
        Self {
            name,
            capture_names: iter.enumerate(),
        }
    }
}

fn main() {
    let r = regex::Regex::new("a").unwrap();
    CaptureIndicies::with_name_and_iter(b"", r.capture_names());
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1456a684a2c5ec1718ca127788c2a6a0

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error[E0433]: failed to resolve: use of undeclared type `CaptureIndicies`
  --> src/main.rs:21:5
   |
21 |     CaptureIndicies::with_name_and_iter(b"", r.capture_names());
   |     ^^^^^^^^^^^^^^^ use of undeclared type `CaptureIndicies`

error[E0412]: cannot find type `CapturesIndicies` in this scope
  --> src/main.rs:10:14
   |
5  | pub struct CaptureIndices<'a, 'b> {
   | --------------------------------- similarly named struct `CaptureIndices` defined here
...
10 | impl<'a, 'b> CapturesIndicies<'a, 'b> {
   |              ^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `CaptureIndices`

error[E0071]: expected struct, variant or union type, found type error
  --> src/main.rs:12:9
   |
12 |         Self {
   |         ^^^^ not a struct

Some errors have detailed explanations: E0071, E0412, E0433.
For more information about an error, try `rustc --explain E0071`.
error: could not compile `playground` due to 3 previous errors
```

<!-- The following is not always necessary. -->
Ideally the output should look like:

the `Self` type not being a struct is caused by the typo in the `impl` target.

Ideally, the diagnostics could recover from "type error" to a suggested fix.

<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

<!--
If dramatically different output is caused by small changes, consider also
adding them here.

If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions. The output might also be
different depending on the Edition.
-->

--- a/compiler/rustc_error_codes/src/error_codes/E0071.md
+++ b/compiler/rustc_error_codes/src/error_codes/E0071.md
@@ -15,13 +15,13 @@ form of initializer was used.
 For example, the code above can be fixed to:
 
 ```
-enum Foo {
-    FirstValue(i32)
-}
+type U32 = u32;
+let t: U32 = 4;
+```
 
-fn main() {
-    let u = Foo::FirstValue(0i32);
+or:
 
-    let t = 4;
-}
+```
+struct U32 { value: u32 }
+let t = U32 { value: 4 };
 ```

--- a/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs
@@ -494,15 +494,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             Some((variant, ty))
         } else {
-            struct_span_err!(
-                self.tcx.sess,
-                path_span,
-                E0071,
-                "expected struct, variant or union type, found {}",
-                ty.sort_string(self.tcx)
-            )
-            .span_label(path_span, "not a struct")
-            .emit();
+            match ty.kind() {
+                ty::Error(_) => {
+                    // E0071 might be caused by a spelling error, which will have
+                    // already caused an error message and probably a suggestion
+                    // elsewhere. Refrain from emitting more unhelpful errors here
+                    // (issue #88844).
+                }
+                _ => {
+                    struct_span_err!(
+                        self.tcx.sess,
+                        path_span,
+                        E0071,
+                        "expected struct, variant or union type, found {}",
+                        ty.sort_string(self.tcx)
+                    )
+                    .span_label(path_span, "not a struct")
+                    .emit();
+                }
+            }
             None
         }
     }

--- a/src/test/ui/typeck/issue-88844.rs
+++ b/src/test/ui/typeck/issue-88844.rs
@@ -1,0 +1,14 @@
+// Regression test for #88844.
+
+struct Struct { value: i32 }
+//~^ NOTE: similarly named struct `Struct` defined here
+
+impl Stuct {
+//~^ ERROR: cannot find type `Stuct` in this scope [E0412]
+//~| HELP: a struct with a similar name exists
+    fn new() -> Self {
+        Self { value: 42 }
+    }
+}
+
+fn main() {}

--- a/src/test/ui/typeck/issue-88844.stderr
+++ b/src/test/ui/typeck/issue-88844.stderr
@@ -1,0 +1,12 @@
+error[E0412]: cannot find type `Stuct` in this scope
+  --> $DIR/issue-88844.rs:6:6
+   |
+LL | struct Struct { value: i32 }
+   | ------------- similarly named struct `Struct` defined here
+...
+LL | impl Stuct {
+   |      ^^^^^ help: a struct with a similar name exists: `Struct`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0412`.

